File: SimpleMatchNums/app/src/main/java/com/example/mjjohnso/simplematch/MainActivity.java
package com.example.mjjohnso.simplematch;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends Activity {

    private static final String SCORE = "SCORE";
    private static final String LAST = "LAST";
    private static final String MATCH = "MATCH";
    private static final String NUMS = "NUMS";

    private TextView mDone;
    private Button mRestart;
    private String mButtonValues[]=new String[16];
    private int mScore=0;
    private int mNumMatched=0;
    private int mLastButtonIndex=-1;
    private Button mButtons[]=new Button[16];
    private final int[] mButtonIds={
            R.id.button11,R.id.button12,R.id.button13,R.id.button14,
            R.id.button21,R.id.button22,R.id.button23,R.id.button24,
            R.id.button31,R.id.button32,R.id.button33,R.id.button34,
            R.id.button41,R.id.button42,R.id.button43,R.id.button44
    };

    private void setButton(final int i,final String s) {
          // mButtons[i].setText(s);

        final int from;
        final int to;
        if (s == "") {
            from = 0;
            to = 180;
        } else {
            from = 180;
            to = 0;
        }
        mButtons[i].setRotationY(from);

        mButtons[i].animate().rotationY((from+to) / 2).setDuration(100).withEndAction(new Runnable() {
            @Override
            public void run() {
                mButtons[i].setText(s);
                mButtons[i].animate().rotationY(to).setDuration(100).withEndAction(new Runnable() {
                    @Override
                    public void run() {
                        mButtons[i].setRotationY(0);
                    }
                });
            }
        });

    }
    @Override
    protected void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putInt(SCORE, mScore);
        savedInstanceState.putInt(LAST, mLastButtonIndex);
        savedInstanceState.putInt(MATCH, mNumMatched);
        savedInstanceState.putStringArray(NUMS, mButtonValues);
        super.onSaveInstanceState(savedInstanceState);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if(savedInstanceState!=null) {
            mScore=savedInstanceState.getInt(SCORE,0);
            mLastButtonIndex=savedInstanceState.getInt(LAST,-1);
            mNumMatched=savedInstanceState.getInt(MATCH,0);
            mButtonValues=savedInstanceState.getStringArray(NUMS);
        }
        mDone=(TextView)findViewById(R.id.done);
        mRestart=(Button)findViewById(R.id.restart);
        mRestart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                init();
            }
        });
        for(int i=0;i<16;i++) {
            final int v=i;
            mButtons[i]=(Button)findViewById(mButtonIds[i]);
            mButtons[i].setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if(!"".equals(mButtons[v].getText()) || mButtons[v].getRotationY()!=0)
                        return;
                    mScore++;
                    setButton(v,mButtonValues[v]);
                    if(mLastButtonIndex==-1) {
                        mLastButtonIndex=v;
                    } else {
                        if(mButtons[mLastButtonIndex].getText().equals(mButtonValues[v])) {
                            mNumMatched++;
                            mLastButtonIndex=-1;
                        } else {
                            setButton(mLastButtonIndex,"");
                            mLastButtonIndex=v;
                        }

                    }
                    showscore();
                }
            });
        }
        if(mScore==0)
            init();
        showscore();
    }

    private void showscore() {
        if(mNumMatched==8) {
            mDone.setText("Complete:"+mScore);
        } else {
            mDone.setText("Score:"+mScore);
        }
    }
    private void init() {
        mNumMatched=0;
        mScore=0;
        mLastButtonIndex=-1;
        for (int i=0;i<16;i++) {
            mButtons[i].setText("");
            mButtonValues[i]="";
        }
        for(int i=1;i<9;i++) {
            int x;
            for(int j=0;j<2;j++) {
                do {
                    x=(int)(Math.random()*16);
                } while(!"".equals(mButtonValues[x]));
                mButtonValues[x]=""+i;
            }
        }
    }
}

File: SimpleMatchNums/app/src/main/java/com/example/mjjohnso/simplematch/MainActivity.kt
package com.example.mjjohnso.simplematch
import android.R.attr.rotation
import android.content.res.Resources
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.AnimationSpec
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ColorScheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

import kotlinx.coroutines.launch

const val ROWS=4
const val COLS=4
const val SCORE = "SCORE"
const val LAST = "LAST"
const val MATCH = "MATCH"
const val NUMS = "NUMS"
const val BUTTONS = "BUTTONS"

const val TAG="MainActivity"

class MainActivity : ComponentActivity() {
    private var score = 0               // how many buttons have been clicked
    private var numMatched = 0          // how many matches are there
    private var lastButtonIndex = -1    // index of the last button clicked

    private var buttonValues = mutableListOf<String>()      // values behind each button
    private var buttonState = mutableStateListOf<String>()  // displayed value of each button
    private var scoreState by mutableStateOf("")    // score displayed on the screen

    // called when a button is clicked
    fun buttonClicked(index: Int) {
        Log.i(TAG, "buttonClicked $index")
        // if the button is already showing a value, do nothing
        if (buttonState[index] == "") {
            score++ // increment score
            buttonState[index] = buttonValues[index] // show the value
            if (lastButtonIndex == -1) { // if this is the first button clicked
                lastButtonIndex = index
            } else { // this is the second button clicked
                // if the two buttons have the same value
                if (buttonState[lastButtonIndex] == buttonValues[index]) {
                    numMatched++ // increment numMatched
                    lastButtonIndex = -1 // keep buttons displayed
                } else { // buttons don't match
                    buttonState[lastButtonIndex] = "" // hide value of last button
                    lastButtonIndex = index // this becomes the last button
                }
            }
        }
        updatescore() // show score
    }
    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(SCORE, score)
        outState.putInt(LAST, lastButtonIndex)
        outState.putInt(MATCH, numMatched)
        outState.putStringArrayList(NUMS, buttonValues.toCollection(ArrayList()))
        val temp = Array(ROWS * COLS) {""}
        for (i in 0..<ROWS * COLS) {
            temp[i] = buttonState[i]
        }
        outState.putStringArray(BUTTONS, temp)
        super.onSaveInstanceState(outState)
    }

    @Composable
    fun MatchGame(modifier: Modifier = Modifier) {
        // Column is passed the modifier so it can add padding to avoid the system UI
        Column(modifier=modifier) {
            Row { // First row shows score and restart button
                Text( // score takes up 3/4 width
                    text = scoreState,
                    modifier = Modifier.weight(3f).align(Alignment.CenterVertically)
                )
                // call init() when the Restart button is pressed
                // restart button takes up 1/4 the width
                Button(onClick = { init() }, modifier = Modifier.weight(1f)) { Text("Restart") }
            }
            // draw the grid of buttons
            for (i in 0..<ROWS)
                Row(modifier = Modifier.weight(1f)) { // each row has equal weight
                    for (j in 0..<COLS) { // draw row of buttons
                        val index = i * COLS + j
                        val rotation by animateFloatAsState(targetValue=if (buttonState[index] == "") 180f else 0f,
                            animationSpec = tween(durationMillis = 1000))
                        Button( // draw a button
                            shape = RoundedCornerShape(4.dp),
                            onClick = { buttonClicked(index) }, // call buttonClicked when pressed
                            modifier = Modifier.weight(1f).fillMaxHeight().padding(4.dp)
                            .graphicsLayer { rotationY=rotation },
                            contentPadding = PaddingValues(0.dp) // remove default padding
                        ) {
                            Text( // draw the button value
                                text = if(rotation<90f) buttonValues[index] else "",
                                fontSize = (6 * 48 / COLS).sp,
                            )
                        }
                    }
                }
        }
    }

    @Preview
    @Composable
    fun ComposablePreview() {
        init()
        MaterialTheme(colorScheme = MaterialTheme.colorScheme) {
            Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
                MatchGame(modifier = Modifier.padding(innerPadding))
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MaterialTheme(colorScheme = MaterialTheme.colorScheme) {
                Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
                    MatchGame(modifier = Modifier.padding(innerPadding))
                }
            }
        }
        init()
        if (savedInstanceState != null) {
            score = savedInstanceState.getInt(SCORE, 0)
            lastButtonIndex = savedInstanceState.getInt(LAST, -1)
            numMatched = savedInstanceState.getInt(MATCH, 0)
            buttonValues = savedInstanceState.getStringArrayList(NUMS)!!
            val temp = savedInstanceState.getStringArray(BUTTONS)!!
            for (i in 0..<ROWS*COLS) {
                buttonState[i]=temp[i]
            }
            updatescore()
        }

    }

    private fun updatescore() {
        scoreState = if (numMatched == ((ROWS * COLS) / 2)) {
            "Complete:$score"
        } else {
            "Score:$score"
        }
    }

    private fun init() {
        numMatched = 0 // initialise score and numMatched
        score = 0
        lastButtonIndex = -1 // no last button
        buttonValues.clear()
        buttonState.clear()
        (1..ROWS*COLS).forEach { // add empty buttons
            buttonValues.add("")
            buttonState.add("")
        }
        for(i in 1..(ROWS * COLS) / 2) {
            var x: Int
            // put pairs of numbers behind random buttons
            (0..1).forEach {
                do {
                    x = (0..(ROWS * COLS)-1).random()
                } while ("" != buttonValues[x])
                buttonValues[x] = "$i"
            }
        }
        updatescore() // set the score
    }
}

